### ISSUE TITLE
Batch size modification not working

### ISSUE THREAD

**Issue by RabbleRousy** [mlpack/mlpack#3110]
I am trying to train a FFN, but it overfits the training data every time, resulting in the prediction being the median of the training labels for every input.

I first used `ens::GradientDescent()` and tried to modify the training parameters like `stepSize`, `maxIterations` and `tolerance`, e.g. by writing:

    network.Train(trainData, trainLabels, ens::GradientDescent(0.1, 128, 0.1), ens::ProgressBar(), ens::PrintLoss());

but none seemed to change the result.

I realized that with my 50k training data, it was always training with `batchSize = 32`, resulting in `50000 / 32 = 1563` batches. I thought maybe increasing the batch size would help to make the network generalize more. Because `ens::GradientDescent` doesn't have a `batchSize` parameter (why?), I tried using `ens::SGD` and `ens::BigBatchSGD`, but with both, changing the batch size did not affect the training at all. The progressbar still looks like this:

    Epoch 1/2
    1/1563 [>...................................................................................................] 0% - ETA:
    2/1563 [>...................................................................................................] 0% - ETA:
    3/1563 [>...................................................................................................] 0% - ETA: 
    4/1563 [>...................................................................................................] 0% - ETA:
    5/1563 [>...................................................................................................] 0% - ETA:
    6/1563 [>...................................................................................................] 0% - ETA:
    7/1563 [>...................................................................................................] 0% - ETA:
    8/1563 [>...................................................................................................] 0% - ETA:
    9/1563 [>...................................................................................................] 0% - ETA:
    10/1563 [>...................................................................................................] 0% - ETA:
    11/1563 [>...................................................................................................] 0% - ETA:
    12/1563 [>...................................................................................................] 0% - ETA:
    13/1563 [>...................................................................................................] 0% - ETA:
    14/1563 [>...................................................................................................] 0% - ETA:
    15/1563 [>...................................................................................................] 0% - ETA:
    16/1563 [=>..................................................................................................] 1% - ETA:
    17/1563 [=>..................................................................................................] 1% - ETA:
    18/1563 [=>..................................................................................................] 1% - ETA:
    19/1563 [=>..................................................................................................] 1% - ETA:
    20/1563 [=>..................................................................................................] 1% - ETA:
    21/1563 [=>..................................................................................................] 1% - ETA:
    22/1563 [=>..................................................................................................] 1% - ETA:
    23/1563 [=>..................................................................................................] 1% - ETA:
    24/1563 [=>..................................................................................................] 1% - ETA:
    25/1563 [=>..................................................................................................] 1% - ETA:
    26/1563 [=>..................................................................................................] 1% - ETA:
    27/1563 [=>..................................................................................................] 1% - ETA:
    28/1563 [=>..................................................................................................] 1% - ETA:
    29/1563 [=>..................................................................................................] 1% - ETA:
    30/1563 [=>..................................................................................................] 1% - ETA:
    31/1563 [=>..................................................................................................] 1% - ETA:
    32/1563 [==>.................................................................................................] 2% - ETA:
    33/1563 [==>.................................................................................................] 2% - ETA:
    34/1563 [==>.................................................................................................] 2% - ETA:
    35/1563 [==>.................................................................................................] 2% - ETA:
    36/1563 [==>.................................................................................................] 2% - ETA:
    37/1563 [==>.................................................................................................] 2% - ETA:
    38/1563 [==>.................................................................................................] 2% - ETA: 
    8s - loss: 0.0204455

Why is the `batchSize` parameter not affecting the training? Might the other parameters also be broken or am I using the wrong syntax for modifying them?

**Comment by RabbleRousy** [comment]
`maxIterations` also seems to be ignored... when training on a 1k dataset and setting `maxIterations = 32`, the training still proceeds to do 100 epochs with 32 batches of size 32 resulting in `100 * 32 * 32 = 102400` iterations ??

**Comment by shrit** [comment]
@RabbleRousy You should set `maxIteration = 0 ` in order for this to work

**Comment by RabbleRousy** [comment]
@shrit you mean like this:

    network.Train(trainData, trainLabels, ens::BigBatchSGD<>(1000, 0.1, 0.1, 0, 0.01), ens::ProgressBar(), ens::PrintLoss());

Doesn't work... (still the same progressbar like above with a 50k dataset)

**Comment by zoq** [comment]
@RabbleRousy what version of ensmallen do you use?

**Comment by RabbleRousy** [comment]
@zoq I cloned ensmallen master on 10.12., so it should be the most recent one.
mlpack version is 3.4.2.

**Comment by zoq** [comment]
I did a quick test with the code below, the logistic function we optimize here has 1000 samples, and if I alter the batch size I can see different results in the progress base output.

Maybe there is something wrong with the way you initialize the function you optimize?

```c++
template<typename MatType>
void LogisticRegressionTestData(MatType& data,
                                   MatType& testData,
                                   MatType& shuffledData, 
                                   arma::Row<size_t>& responses,
                                   arma::Row<size_t>& testResponses,
                                   arma::Row<size_t>& shuffledResponses) 
{
  // Generate a two-Gaussian dataset.
  data = MatType(3, 1000);
  responses = arma::Row<size_t>(1000);
  for (size_t i = 0; i < 500; ++i)
  {
    // The first Gaussian is centered at (1, 1, 1) and has covariance I.
    data.col(i) = arma::randn<arma::Col<typename MatType::elem_type>>(3) 
        arma::Col<typename MatType::elem_type>("1.0 1.0 1.0");
    responses(i) = 0;
  }
  for (size_t i = 500; i < 1000; ++i)
  {
    // The second Gaussian is centered at (9, 9, 9) and has covariance I.
    data.col(i) = arma::randn<arma::Col<typename MatType::elem_type>>(3) 
        arma::Col<typename MatType::elem_type>("9.0 9.0 9.0");
    responses(i) = 1;
  }

  // Shuffle the dataset.
  arma::uvec indices = arma::shuffle(arma::linspace<arma::uvec>(0, 
      data.n_cols - 1, data.n_cols));
  shuffledData = MatType(3, 1000);
  shuffledResponses = arma::Row<size_t>(1000);
  for (size_t i = 0; i < data.n_cols; ++i)
  {
    shuffledData.col(i) = data.col(indices(i));
    shuffledResponses(i) = responses[indices(i)];
  }

  // Create a test set.
  testData = MatType(3, 1000);
  testResponses = arma::Row<size_t>(1000);
  for (size_t i = 0; i < 500; ++i)
  {
    testData.col(i) = arma::randn<arma::Col<typename MatType::elem_type>>
        arma::Col<typename MatType::elem_type>("1.0 1.0 1.0");
    testResponses(i) = 0;
  }
  for (size_t i = 500; i < 1000; ++i)
  {
    testData.col(i) = arma::randn<arma::Col<typename MatType::elem_type>>
        arma::Col<typename MatType::elem_type>("9.0 9.0 9.0");
    testResponses(i) = 1;
  }
}

AMSGrad amsgrad(1e-3, 4, 0.9, 0.999, 1e-8, 500000, -10, true);
amsgrad.MaxIterations() = 100000;
/* NumFunctions = 1000 */

arma::mat data, testData, shuffledData;
arma::Row<size_t> responses, testResponses, shuffledResponses;

LogisticRegressionTestData(data, testData, shuffledData,
    responses, testResponses, shuffledResponses);
ens::test::LogisticRegression<arma::mat> lr(shuffledData, shuffledResponses,
    0.5);

arma::mat coordinates = lr.GetInitialPoint();

amsgrad.Optimize(lr, coordinates, ens::ProgressBar());
```


Output for `BatchSize = 4 (1000 / 4) = 250`:

```
Epoch 1/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.480714
Epoch 2/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.39607
Epoch 3/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.350897
Epoch 4/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.31227
Epoch 5/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.278578
Epoch 6/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.249518
Epoch 7/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.224251
Epoch 8/100
250/250 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.202333
```

Output for `BatchSize = 4 (1000 / 3) = 333.33333`

```
Epoch 1/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.473476
Epoch 2/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.382269
Epoch 3/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.329272
Epoch 4/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.28505
Epoch 5/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.247649
Epoch 6/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.216738
Epoch 7/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.191129
Epoch 8/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.169883
Epoch 9/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.15234
Epoch 10/100
334/334 [====================================================================================================] 100% - 0s 0ms/step - loss: 0.137709
```
Or maybe I'm missing something?

**Comment by RabbleRousy** [comment]
I gotta admit I don't really understand a lot of what's going on in your example, but what immediately caught my eye was that you were using the `Optimize()` function, while I am trying to use `FFN<>::Train()` and pass the optimizer as a parameter like this:

    mlpack::ann::FFN<mlpack::ann::MeanSquaredError<>> network;

	// LinearSplit and ClippedReLU written by me, see previous issue
	network.Add<mlpack::ann::LinearSplit<> >(2 * N, 2 * M);
	network.Add<mlpack::ann::ClippedReLULayer<>>();
								 
	network.Add<mlpack::ann::Linear<> >(2 * M, K);
	network.Add<mlpack::ann::ClippedReLULayer<>>();
								 
	network.Add<mlpack::ann::Linear<> >(K, K);
	network.Add<mlpack::ann::ClippedReLULayer<>>();

	network.Add<mlpack::ann::Linear<> >(K, 1);

    network.Train(trainData, trainLabels, ens::GradientDescent(0.1, 128, 0.1), ens::PrintLoss(), ens::ProgressBar());

Is there a way to replace the last line with a call to `Optimize()`? Could be worth a try...

Anyway, I will come back to this after christmas. Thanks a lot for all your help already, especially @shrit & @zoq !
Happy Holidays! :)

**Comment by RabbleRousy** [comment]
I tried using `AMSGrad` like in @zoq 's example, and **it works**!
I can modify batchSize, stepSize, all the good stuff and the training results already look waaay better :)
So maybe there's just something wrong with the other classes and they are not really compatible anymore?
Thanks again for all your help!
